---
title: "Gene set analysis misuse examples"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 7
    fig_height: 7
theme: cosmo
---

Source: https://github.com/markziemann/combined_enrichment

## Intro

This script is designed to aggregate the results from seven RNA-seq enrichment analyses.

The purpose here is to get some idea as to how consistent the results are across several datasets.

Data was analysed with mitch (representing FCS) and clusterprofiler (representing ORA).

Here are the datasets considered.

| Dataset | SRA accesion | genes in annotation set | genes detected | genes differentially expressed |
| :---: | :---: | :---: | :---: | :---: | 
| 1. | SRP128998 | 39297 | 15635 | 3472 |
| 2. | SRP038101 | 39297 | 13926 | 3589 | 
| 3. | SRP096178 | 39297 | 15477 | 9488 |
| 4. | SRP038101 | 39297 | 15607 | 5150 | 
| 5. | SRP247621 | 39297 | 14288 | 230 |
| 6. | SRP253951 | 39297 | 15182 | 8588 |
| 7. | SRP068733 | 39297 | 14255 | 7365 |

```{r,begin0}

suppressPackageStartupMessages({
library("kableExtra")
library("beeswarm")
})

```

## Execute the examples

Here I'm running the scripts to analyse each of the example datasets shown in the above
table.

Only execute the examples if file `ex7dat.rds` does not exist.

```{r,execute0}

if ( ! file.exists("ex7dat.rds") ) {
  rmarkdown::render("example1.Rmd")
  rmarkdown::render("example2.Rmd")
  rmarkdown::render("example3.Rmd")
  rmarkdown::render("example4.Rmd")
  rmarkdown::render("example5.Rmd")
  rmarkdown::render("example6.Rmd")
  rmarkdown::render("example7.Rmd")
}

```

## Summarise

Here I'm summarising the results of the five example data sets.
Let's start with FCS and then ORA

```{r,getdata0,fig.width=7,fig.height=5}

exdat=NULL
ex1dat <- readRDS("ex1dat.rds")
ex2dat <- readRDS("ex2dat.rds")
ex3dat <- readRDS("ex3dat.rds")
ex4dat <- readRDS("ex4dat.rds")
ex5dat <- readRDS("ex5dat.rds")
ex6dat <- readRDS("ex6dat.rds")
ex7dat <- readRDS("ex7dat.rds")

msummary <- function(dat) {
  m_all  <- unique(c(dat$FCS_up ,dat$FCS_dn , dat$FCS_com ))
  m_sep_rec <- length(union(dat$FCS_up, dat$FCS_dn)) / length(m_all)
  m_com_rec <- length(dat$FCS_com) / length(m_all)
  res <- c("FCS up"=length(dat$FCS_up),
    "FCS dn"=length(dat$FCS_dn),
    "FCS sep"=length(union(dat$FCS_up,dat$FCS_dn)),
    "FCS com"=length(dat$FCS_com),
    "FCS all"=length(m_all),
    "FCS sep rec"=m_sep_rec,
    "FCS com rec"=m_com_rec,
    "FCS sep only"=length(setdiff(union(dat$FCS_up,dat$FCS_dn),dat$FCS_com )),
    "FCS com only"=length(setdiff(dat$FCS_com , union(dat$FCS_up,dat$FCS_dn))) )
  return(res)
}

mres <- lapply(list(ex1dat,ex2dat,ex3dat,ex4dat,ex5dat,ex6dat,ex7dat),msummary)
mres <- do.call(rbind,mres)
rownames(mres) <- c("SRP128998","SRP038101","SRP096178","SRP038101","SRP247621","SRP253951","SRP068733")
mres %>% kbl(caption="differentially expressed pathway detection using separated or combined FCS analysis for seven studies") %>% kable_paper("hover", full_width = F)
# list format
mresl <- lapply(1:ncol(mres),function(i) {mres[,i]} )
names(mresl) <- colnames(mres)

osummary <- function(dat) {
  ORA_all  <- unique(c(dat$ORA_up ,dat$ORA_dn , dat$ORA_com ))
  ORA_sep_rec <- length(union(dat$ORA_up, dat$ORA_dn)) / length(ORA_all)
  ORA_com_rec <- length(dat$ORA_com) / length(ORA_all)
  res <- c("ORA up"=length(dat$ORA_up),
    "ORA dn"=length(dat$ORA_dn),
    "ORA sep"=length(union(dat$ORA_up,dat$ORA_dn)),
    "ORA com"=length(dat$ORA_com),
    "ORA all"=length(ORA_all),
    "ORA sep rec"=ORA_sep_rec,
    "ORA com rec"=ORA_com_rec,
    "ORA sep only"=length(setdiff(union(dat$ORA_up,dat$ORA_dn),dat$ORA_com )),
    "ORA com only"=length(setdiff(dat$ORA_com , union(dat$ORA_up,dat$ORA_dn))) )
  return(res)
}

ores <- lapply(list(ex1dat,ex2dat,ex3dat,ex4dat,ex5dat,ex6dat,ex7dat),osummary)
ores <- do.call(rbind,ores)
rownames(ores) <- c("SRP128998","SRP038101","SRP096178","SRP038101","SRP247621","SRP253951","SRP068733")
ores %>% kbl(caption="differentially expressed pathway detection using separated or combined ORA analysis for seven studies") %>% kable_paper("hover", full_width = F)
# list format
oresl <- lapply(1:ncol(ores),function(i) {ores[,i]} )
names(oresl) <- colnames(ores)

```

```{r,summary_stats0}

means <- colMeans(mres)
medians <- apply(mres,2,median)
sds <- apply(mres,2,sd)
summstats <- rbind(means,medians,sds)
summstats %>% kbl(caption="summary statistics of FCS results") %>% kable_paper("hover", full_width = F)


means <- colMeans(ores)
medians <- apply(ores,2,median)
sds <- apply(ores,2,sd)
summstats <- rbind(means,medians,sds)
summstats %>% kbl(caption="summary statistics of ORA results") %>% kable_paper("hover", full_width = F)


```

Here let's plot the jaccard scores.
First as a boxplot and then overlay the dataset studies over the top.

Mitch results.

```{r,jacbox_fcs}

cols <- c("#CC79A7","#56B4E9","#E69F00","#F0E442","#009E73","#0072B2","#D55E00")
coll <- list(cols, cols, cols, cols, cols)
par(mar=c(5,5,3,1))
boxplot(mres[,1:5],col="white",names = gsub("FCS ","",colnames(mres)[1:5]),ylab="no. gene sets identified",main="FCS")
beeswarm(mresl[1:5] , pwcol = coll, cex=1.5,pch=19, add = TRUE, main="FCS")
legend("topleft", legend=c("SRP128998", "SRP038101", "SRP096178","SRP038101","SRP247621","SRP253951","SRP068733"), pch=19,  col=cols, cex=1,pt.cex = 1.5)

coll <- list(cols, cols)
mynames <- sub(" rec","",gsub("FCS ","",colnames(mres)[6:7]))
boxplot(mres[,6:7],col="white", names = mynames, ylab="proportion of gene sets identified",main="FCS")
beeswarm(mresl[6:7] , pwcol = coll, cex=1.5,pch=19, add = TRUE)
legend("bottomleft", legend=c("SRP128998", "SRP038101", "SRP096178","SRP038101","SRP247621","SRP253951","SRP068733"), pch=19,  col=cols, cex=1,pt.cex = 1.5)

coll <- list(cols, cols)
mynames <- sub(" rec","",gsub("FCS ","",colnames(mres)[8:9]))
boxplot(mres[,8:9],col="white", names = mynames, ylab="no. gene sets unique to each approach",main="FCS")
beeswarm(mresl[8:9] , pwcol = coll, cex=1.5,pch=19, add = TRUE)
legend("topright", legend=c("SRP128998", "SRP038101", "SRP096178","SRP038101","SRP247621","SRP253951","SRP068733"), pch=19,  col=cols, cex=1,pt.cex = 1.5)

```

ORA results.

```{r,jacbox_ora}

cols <- c("#CC79A7","#56B4E9","#E69F00","#F0E442","#009E73","#0072B2","#D55E00")
coll <- list(cols, cols, cols, cols, cols)
par(mar=c(5,5,3,1))
boxplot(ores[,1:5],col="white",names = gsub("ORA ","",colnames(ores)[1:5]),ylab="no. gene sets identified",main="ORA")
beeswarm(oresl[1:5] , pwcol = coll, cex=1.5,pch=19, add = TRUE, main="ORA")
legend("topleft", legend=c("SRP128998", "SRP038101", "SRP096178","SRP038101","SRP247621","SRP253951","SRP068733"), pch=19,  col=cols, cex=1,pt.cex = 1.5)

coll <- list(cols, cols)
mynames <- sub(" rec","",gsub("ORA ","",colnames(ores)[6:7]))
boxplot(ores[,6:7],col="white", names = mynames, ylab="proportion of gene sets identified",main="ORA")
beeswarm(oresl[6:7] , pwcol = coll, cex=1.5,pch=19, add = TRUE)
legend("bottomleft", legend=c("SRP128998", "SRP038101", "SRP096178","SRP038101","SRP247621","SRP253951","SRP068733"), pch=19,  col=cols, cex=1,pt.cex = 1.5)

coll <- list(cols, cols)
mynames <- sub(" rec","",gsub("ORA ","",colnames(ores)[8:9]))
boxplot(ores[,8:9],col="white", names = mynames, ylab="no. gene sets unique to each approach",main="ORA")
beeswarm(oresl[8:9] , pwcol = coll, cex=1.5,pch=19, add = TRUE)
legend("topright", legend=c("SRP128998", "SRP038101", "SRP096178","SRP038101","SRP247621","SRP253951","SRP068733"), pch=19,  col=cols, cex=1,pt.cex = 1.5)

```

## Conclusions






## Session information

```{r,session0}

sessionInfo()

```
